### PR TITLE
Query parser: encode ?? as ?

### LINKED ISSUE #60: Variable binding fails if `?` appears within a string in query sql.
It seems that at least some rudimentary parsing should be done in order to only respond to `?` that are not parts of string literals.

As a kludge we could alternatively have a method that says to ignore any `?` that are present.

### PR DESCRIPTION
**Pull request by serprex**
fixes #60

### CODE REVIEW

**Comment by serprex** [comment]
@cvincent-reciprocity would this address your problem?

**Comment by serprex** [comment]
An opt out constructor could be made which takes the input string & puts it into `vec![Part::Str(input)]`

**Comment by cvincent-reciprocity** [comment]
> @cvincent-reciprocity would this address your problem?

Yes!

File: src/sql/mod.rs [hunk 1]
```diff
@@ -12,34 +12,41 @@ pub(crate) mod escape;
...
+    pub(crate) fn new(template: &'parts str) -> Self {
+        let mut parts = Vec::new();
+        let mut rest = template;
+        while let Some(idx) = rest.find('?') {
+            if idx != 0 {
+                parts.push(Part::Str(&rest[..idx]));
+            }
+            rest = &rest[idx + 1..];
+            if let Some(restqq) = rest.strip_prefix('?') {
+                parts.push(Part::Str("?"));
```
Comment (serprex) [inline src/sql/mod.rs:38]: Instead of being a separate Part, this could have the above part slice to idx+1 instead of idx

File: src/sql/mod.rs [hunk 2]
```diff
@@ -12,34 +12,43 @@ pub(crate) mod escape;
...
-    pub(crate) fn new(template: &str) -> Self {
-        let mut iter = template.split('?');
-        let prefix = String::from(iter.next().unwrap());
-        let mut parts = vec![Part::Text(prefix)];
+impl<'parts> SqlBuilder<'parts> {
+    pub(crate) fn new(template: &'parts str) -> Self {
+        let mut parts = Vec::new();
+        let mut rest = template;
+        while let Some(idx) = rest.find('?') {
+            if rest[idx + 1..].starts_with('?') {
```
Comment (pravic) [inline src/sql/mod.rs:33]: What if we do `client.query("?")`? Or panic is [okay](https://github.com/ClickHouse/clickhouse-rs/issues/60#issuecomment-1546850356)?

Comment (serprex) [reply src/sql/mod.rs:33]: for `client.query("?")` `rest[idx + 1..]` here will be `""`, & `starts_with` will return `false`

Comment (pravic) [reply src/sql/mod.rs:33]: You are right, sorry.

File: src/lib.rs [hunk 3]
```diff
@@ -299,7 +299,7 @@ impl Client {
     }
 
     /// Starts a new SELECT/DDL query.
-    pub fn query(&self, query: &str) -> query::Query {
+    pub fn query<'a>(&self, query: &'a str) -> query::Query<'a> {
```
Comment (loyd) [inline src/lib.rs:302]: Avoiding allocations here is not worth introducing such a breaking change in public API.

* Allocations are (usually) done in the same thread as deallocations, so for most modern allocators it's cheap (any performance-focused code anyway will use mimalloc/tcmalloc/jemalloc)
* Allocations here are rare operations compared to decoding incoming rows
* Constructing a query is negligible compared to HTTP and CH

`Query`/`Watch` can be stored and even sent; I'm not sure we should prevent a such usage.

File: src/sql/mod.rs [hunk 4]
```diff
@@ -12,34 +12,43 @@ pub(crate) mod escape;
...
+    Str(&'a str),
+    String(String),
 }
 
-impl SqlBuilder {
-    pub(crate) fn new(template: &str) -> Self {
-        let mut iter = template.split('?');
-        let prefix = String::from(iter.next().unwrap());
-        let mut parts = vec![Part::Text(prefix)];
+impl<'parts> SqlBuilder<'parts> {
```
Comment (loyd) [inline src/sql/mod.rs:28]: This code obviously requires unit tests

File: src/sql/mod.rs [hunk 5]
```diff
@@ -57,7 +66,7 @@ impl SqlBuilder {
                 return self.error(format_args!("invalid argument: {err}"));
             }
 
-            *part = Part::Text(s);
+            *part = Part::String(s);
```
[the remark this answers is not in the document]
Comment (loyd) [reply src/sql/mod.rs:69]: I don't think we should optimize this part of the crate for now. Do you have a real case where it matters?

**Review by loyd — CHANGES_REQUESTED** [review]

File: src/sql/mod.rs [hunk 6]
```diff
@@ -21,25 +21,34 @@ pub(crate) enum SqlBuilder {
 pub(crate) enum Part {
     Arg,
     Fields,
+    Str(&'static str),
     Text(String),
 }
 
 impl SqlBuilder {
     pub(crate) fn new(template: &str) -> Self {
```
Comment (serprex) [inline src/sql/mod.rs:29]: with removal of lifetime generics, a constructor could still be added for when query is `&'static str`

Anyways, that'd be a different PR

File: src/sql/mod.rs [hunk 7]
```diff
@@ -184,6 +190,15 @@ mod tests {
         );
     }
 
+    #[test]
```
Comment (loyd) [inline src/sql/mod.rs:193]: Need to note about `??` in `bind()` doc

Comment (serprex) [reply src/sql/mod.rs:193]: added mention in warning

**Review by loyd — APPROVED** [review]
LGTM

**Comment by loyd** [comment]
@serprex, can you rebase (easy to squash firstly) the PR?

**Pull request merged**